contracts/gol.cairo

```diff
@@ -0,0 +1,450 @@
+func view_game{
+        syscall_ptr : felt*,
+        bitwise_ptr : BitwiseBuiltin*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        game_id : felt,
+        generation : felt
+    ) -> (
+        generation_owner : felt,
+        game_state : felt
+    ):
+
+    let (game_info) = stored_game.read(game_id, generation)
+    return (game_info[0], game_info[1])
+end
+
+# Returns the current generation of a given game.
+@view
+func get_current_generation{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        game_id : felt
+    ) -> (
+        generation : felt
+    ):
+    let (generation) = current_generation.read(game_id)
+    return (generation)
+end
+
+##########
+
+func pay{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        user : felt, 
+        credit_requirement : felt
+    ):
+    ERC20._burn(user, Uint256(credit_requirement, 0))
+    return ()
+end
+
+func reward_user{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        user : felt
+    ):
+    ERC20._mint(user, Uint256(1, 0))
+    return ()
+end
+
+func ensure_user{
+        syscall_ptr : felt*
+    }(
+    ) -> (
+        caller : felt
+    ):
+    let (caller) = get_caller_address()
+    with_attr error_message(
+        "User not authenticated"
+    ):
+        assert_not_zero(caller)
+    end
+    return (caller)
+end
+
+func verify_new_game{
+        syscall_ptr : felt*,
+        bitwise_ptr : BitwiseBuiltin*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        game : felt
+    ):
+    alloc_locals
+    let (local game_info) = stored_game.read(
+        game_id=game, generation=0)
+    let game_id = game_info[1]
+
+    with_attr error_message(
+        "Game with genesis state {game} already exists"
+    ):
+        assert_not_equal(game_id, game)
+    end
+
+    let (high, _) = split_felt(game)
+    let (invalid_bits, _) = unsigned_div_rem(high, 2**97)
+    with_attr error_message(
+        "Game {game} is not a valid game"
+    ):
+        assert invalid_bits = 0
+    end
+    return ()
+end
+
+func verify_and_get_for_cell_activation{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        cell_index : felt,
+        user : felt
+    ) -> (
+        current_generation : felt,
+        current_state : felt
+    ):
+    alloc_locals
+    let (generation) = current_generation.read(
+        game_id=INFINITE_GAME_GENESIS
+    )
+    let (local game_info) = stored_game.read(
+        game_id=INFINITE_GAME_GENESIS,
+        generation=generation
+    )
+    let current_game_state = game_info[1]
+    with_attr error_message(
+        "Cell index {cell_index} out of range"
+    ):
+        assert_le_felt(cell_index, DIM*DIM-1)
+    end
+    return(generation, current_game_state)
+end
+
+func save_game{
```

Comment:
This is actually `create_new_game`, generation is not needed as a param, it is always 0.